### PR TITLE
Refactoring Dropdown props

### PR DESCRIPTION
**Pull request by Zombiefruit**

### CODE REVIEW

File: src/components/drop-down/drop-down.tsx [hunk 1]
```diff
@@ -1,66 +1,58 @@
...
+        value: 'Default Text',
+        children: [],
+        onChange: noop,
+        tabIndex: 0
+    };
 
-    public static defaultProps: DropDownProps = {items: [], onItemClick: () => {}, onInputClick: () => {}};
+    constructor() {
+        super();
+        this.state = {dropdown: null};
```
Comment (liorzisman) [inline src/components/drop-down/drop-down.tsx]: can replace constructor with `public state = {dropdown: null};`

File: src/components/drop-down/drop-down.tsx [hunk 2]
```diff
@@ -1,66 +1,58 @@
...
-
-export interface DropDownItem {
-    label: string;
+export interface DropDownProps extends OptionProps {
+    open?: boolean;
+    value?: string;
+    onChange?: (id: string) => void;
+    disabled?: boolean;
+    openOnFocus?: boolean;
+    children?: any;
```
Comment (liorzisman) [inline src/components/drop-down/drop-down.tsx]: any?

File: src/components/drop-down/drop-down.tsx [hunk 3]
```diff
@@ -70,15 +62,62 @@ export class DropDown extends React.Component<DropDownProps, {}> {
...
+                    }
+                </div>
+                <Popup open={!!this.props.open && !this.props.disabled} anchor={this.state.dropdown}>
+                    <div className="root">
+                        <SelectionList
+                            data-automation-id="DROP_DOWN_LIST"
+                            className="drop-down-list"
+                            value={this.props.value}
+                            onChange={this.onItemClick!}
+                            dataSource={this.props.children}
```
Comment (liorzisman) [inline src/components/drop-down/drop-down.tsx]: is children a good name for it? why not dataSource? its also extended from the OptionList interface

File: test/components/drop-down.spec.tsx [hunk 4]
```diff
@@ -45,46 +51,106 @@ describe('<DropDown />', () => {
     });
 
     it('has correct selected item text', () => {
-        const item = {label: 'Test'};
-        const {select, waitForDom} = clientRenderer.render(<DropDown selectedItem={item}/>);
+        const item = 'Test';
+        const {select, waitForDom} = clientRenderer.render(<DropDown value={item}>{item}</DropDown>);
```
Comment (liorzisman) [inline test/components/drop-down.spec.tsx]: you render the item inside the dropdown?

File: test/components/drop-down.spec.tsx [hunk 5]
```diff
@@ -45,46 +51,106 @@ describe('<DropDown />', () => {
...
+        await waitForDom(() => expect(onChange).to.have.been.calledOnce);
+    });
+
     it('invokes the onClick when dropdown label is clicked', () => {
         const onClick = sinon.spy();
-        const {select} = clientRenderer.render(<DropDown onInputClick={onClick}/>);
+        const {select} = clientRenderer.render(<DropDown onChange={onClick}/>);
         simulate.click(select(dropDown, input));
 
         return waitFor(() => expect(onClick).to.have.been.calledOnce);
```
Comment (liorzisman) [inline test/components/drop-down.spec.tsx]: can we assert that the event has been invoked with correct data?

File: test/components/drop-down.spec.tsx [hunk 6]
```diff
@@ -45,46 +51,106 @@ describe('<DropDown />', () => {
...
+    it('calls onChange when focused and openOnFocus is true', async () => {
+        const onChange = sinon.spy();
+        const {select, waitForDom} = clientRenderer.render(<DropDown openOnFocus onChange={onChange} />);
+
+        simulate.focus(select(dropDown));
+
+        await waitForDom(() => expect(onChange).to.have.been.calledOnce);
+    });
+
     it('invokes the onClick when dropdown label is clicked', () => {
```
Comment (liorzisman) [inline test/components/drop-down.spec.tsx]: can we assert that the event has been invoked with correct data?

File: test/components/drop-down.spec.tsx [hunk 7]
```diff
@@ -45,46 +51,106 @@ describe('<DropDown />', () => {
...
 
         return waitFor(() => expect(onClick).to.have.been.calledOnce);
     });
 
     it('displays item list to choose from when open is true', () => {
-        const {select, waitForDom} =
-            clientRenderer.render(<DropDown selectedItem={undefined} open={true} items={items} />);
-        const dropDownList = select(dropDown, list, 'LIST');
+        const {waitForDom} =
+            clientRenderer.render(<DropDown value={undefined} open={true} children={items} />);
```
Comment (liorzisman) [inline test/components/drop-down.spec.tsx]: is value `undefined` the default?

File: src/components/drop-down/drop-down.tsx [hunk 8]
```diff
@@ -70,15 +62,62 @@ export class DropDown extends React.Component<DropDownProps, {}> {
...
-                    items={this.props.items}
-                    open={!!this.props.open}
-                    onItemClick={this.onItemClick}
-                />
+            <div
+                {...rootProps}
+                onKeyDown={this.onKeyDown}
+                onFocus={this.onFocus}
+                tabIndex={this.props.tabIndex}
+                ref={dropdown => {
```
Comment (liorzisman) [inline src/components/drop-down/drop-down.tsx]: this format always keeps ref to the first dom node you had. there is no need for the if condition

File: src/components/drop-down/drop-down.tsx [hunk 9]
```diff
@@ -70,15 +62,62 @@ export class DropDown extends React.Component<DropDownProps, {}> {
...
+            >
+                <div data-automation-id="DROP_DOWN_INPUT" onClick={this.onInputClick} className="drop-down-input">
+                    <span className="label">{this.props.value!}</span>
+                    {this.props.toggleIcon
+                        ? this.props.toggleIcon
+                        : <CaretDown className="caret" />
+                    }
+                </div>
+                <Popup open={!!this.props.open && !this.props.disabled} anchor={this.state.dropdown}>
+                    <div className="root">
```
Comment (liorzisman) [inline src/components/drop-down/drop-down.tsx]: why another div to wrap the list?

File: src/components/drop-down/drop-down.tsx (hunk 3 above)
Comment (liorzisman) [inline src/components/drop-down/drop-down.tsx]: `dataSource={this.props.children}` makes no sense. what is the type of dataSource?

File: src/components/drop-down/drop-down.tsx [hunk 10]
```diff
@@ -69,16 +57,60 @@ export class DropDown extends React.Component<DropDownProps, {}> {
...
+                    >
+                        {this.props.children}
+                    </SelectionList>
+                </Popup>
             </div>
         );
     }
+
+    private onFocus = () => {
+        if (this.props.openOnFocus) { this.toggleDropDown(); }
```
Comment (tobich) [inline src/components/drop-down/drop-down.tsx]: pls no single line blocks. Either without `{` or multiline

File: src/components/drop-down/drop-down.tsx [hunk 11]
```diff
@@ -69,16 +57,60 @@ export class DropDown extends React.Component<DropDownProps, {}> {
...
+    }
+
+    private toggleDropDown() {
+        if (this.props.open) { this.state.dropdown!.focus(); }
```
Comment (tobich) [inline src/components/drop-down/drop-down.tsx]: dtto

File: src/components/drop-down/drop-down.tsx [hunk 12]
```diff
@@ -69,16 +57,60 @@ export class DropDown extends React.Component<DropDownProps, {}> {
...
+                tabIndex={this.props.tabIndex}
+                ref={dropdown => this.setState({dropdown})}
+            >
+                <div data-automation-id="DROP_DOWN_INPUT" onClick={this.onInputClick} className="drop-down-input">
+                    <span className="label">{this.props.value!}</span>
+                    <div className="caret" data-automation-id="ICON">
+                        <ToggleIcon />
+                    </div>
+                </div>
+                <Popup open={!!this.props.open && !this.props.disabled} anchor={this.state.dropdown}>
```
Comment (tobich) [inline src/components/drop-down/drop-down.tsx]: no need for !!

File: src/components/drop-down/drop-down.tsx [hunk 13]
```diff
@@ -1,66 +1,54 @@
...
+        toggleIcon: CaretDown
+    };
 
-    public static defaultProps: DropDownProps = {items: [], onItemClick: () => {}, onInputClick: () => {}};
+    public state: DropDownState = {dropdown: null};
 
     public onItemClick = (item: string) => {
-        this.props.onInputClick!();
-        this.props.onItemClick!(this.props.items!.filter((elem: DropDownItem) => elem.label === item)[0]);
+        this.toggleDropDown();
```
Comment (tobich) [inline src/components/drop-down/drop-down.tsx]: Should close drop down, rather than toggle, no?

File: src/components/drop-down/drop-down.tsx [hunk 14]
```diff
@@ -69,16 +57,60 @@ export class DropDown extends React.Component<DropDownProps, {}> {
...
-                    onItemClick={this.onItemClick}
-                />
+            <div
+                {...rootProps}
+                onKeyDown={this.onKeyDown}
+                onFocus={this.onFocus}
+                tabIndex={this.props.tabIndex}
+                ref={dropdown => this.setState({dropdown})}
+            >
+                <div data-automation-id="DROP_DOWN_INPUT" onClick={this.onInputClick} className="drop-down-input">
```
Comment (tobich) [inline src/components/drop-down/drop-down.tsx]: Shouldn't onClick={this.onInputClick} be onClick={this.toggle} ?

File: test/components/drop-down.spec.tsx [hunk 15]
```diff
@@ -1,37 +1,43 @@
...
 import {ClientRenderer, expect, simulate, sinon, waitFor} from 'test-drive-react';
 import {DropDownDemo} from '../../demo/components/drop-down.demo';
 import {DropDown} from '../../src';
 
 const dropDown = 'DROP_DOWN';
 const dropDownDemo = dropDown + '_DEMO';
 const input = dropDown + '_INPUT';
-const list = dropDown + '_LIST';
+
+const KeyCodes: any = {
```
Comment (tobich) [inline test/components/drop-down.spec.tsx]: This is a code duplicate, innit?

File: src/components/drop-down/drop-down.tsx [hunk 16]
```diff
@@ -70,16 +60,77 @@ export class DropDown extends React.Component<DropDownProps, {}> {
...
+        }
+    }
+
+    private closeDropdown() {
+        if (!this.props.disabled) {
+            this.setState({open: false});
+        }
+    }
+
+    private onKeyDown = (e: any) => {
```
Comment (tobich) [inline src/components/drop-down/drop-down.tsx]: why 'any'?

File: src/components/drop-down/drop-down.tsx [hunk 17]
```diff
@@ -1,67 +1,57 @@
...
-}, style);
-
-export interface DropDownItem {
-    label: string;
+export interface DropDownProps extends OptionList, FormInputProps<string> {
+    open?: boolean;
+    disabled?: boolean;
+    openOnFocus?: boolean;
+    children?: React.ReactNode;
+    toggleIcon?: React.SFC;
```
Comment (tobich) [inline src/components/drop-down/drop-down.tsx]: No need to limit it just to stateless components. Its should be `React.ComponentType<>`

File: src/components/drop-down/drop-down.tsx (hunk 16 above)
Comment (Zombiefruit) [reply src/components/drop-down/drop-down.tsx]: Not sure, I'll change it to be more specific

**Pull request merged**